### PR TITLE
feat: Add Reddit ingest connector

### LINKED ISSUE #272: Create a data connector for Reddit
Create a data connector that:
* fetches one more Reddit messages from a Subreddit or Redditor
* stores them locally as markdown files (at least temporarily for processing)
* processes the files the standard way, through there is support for them through [unstructured.partition.auto](https://github.com/Unstructured-IO/unstructured/blob/main/unstructured/partition/auto.py).

See [Adding Data Connectors](https://github.com/Unstructured-IO/unstructured/blob/main/Ingest.md#adding-data-connectors) for details on how to get started. Make sure to include a link to this issue when submitting a PR.

For inspiration on downloading Reddit content, see:
https://github.com/emptycrown/llama-hub/blob/main/loader_hub/reddit/base.py

**Additional References**
* [PRAW: The Python Reddit API Wrapper](https://praw.readthedocs.io/en/stable/)
* https://www.reddit.com/dev/api/

**Definition of Done**

* The [checklist](https://github.com/Unstructured-IO/unstructured/blob/main/Ingest.md#the-checklist) has been completed.
* The connector is able to process all documents for a subreddit or user, at least up to a large well defined limit
* May constrain messages processed by date range

### PR DESCRIPTION
**Pull request by tomaarsen**
Closes #272

Hello!

## Pull Request overview
* Add Reddit data connector for ingest.
  * I have ticked off **most** of the checklist.
  * The connector can process a subreddit.
    * Either via a search query,
    * or via hot posts.
  * The texts in the submissions are converted to markdown files including the post title and the text body, if any (i.e. no images or videos).
  * The number of posts to fetch can be changed with the CLI.

## The data connector checklist
- [x] Create a new module under [unstructured/ingest/connector/](unstructured/ingest/connector/) implementing the 3 abstract base classes, similar to [unstructured/ingest/connector/s3_connector.py](unstructured/ingest/connector/s3_connector.py).
  - [x] The subclass of `BaseIngestDoc` overrides `process_file()` if extra processing logic is needed other than what is provided by [auto.partition()](unstructured/partition/auto.py).
- [x] Update [unstructured/ingest/main.py](unstructured/ingest/main.py) with support for the new connector.
- [x] Create a folder under [examples/ingest](examples/ingest) that includes at least one well documented script.
- [ ] Add a script test_unstructured_ingest/test-ingest-\<the-new-data-source\>.sh. It's json output files should have a total of no more than 100K.
- [ ] Git add the expected outputs under test_unstructured_ingest/expected-structured-output/\<folder-name-relevant-to-your-dataset\> so the above test passes in CI.
- [ ] Add a line to [test_unstructured_ingest/test-ingest.sh](test_unstructured_ingest/test-ingest.sh) invoking the new test script.
- [x] If additional python dependencies are needed for the new connector:
  - [x] Add them as an extra to [setup.py](unstructured/setup.py).
  - [x] Update the Makefile, adding a target for `install-ingest-<name>` and adding another `pip-compile` line to the `pip-compile` make target. See [this commit](https://github.com/Unstructured-IO/unstructured/commit/ab542ca3c6274f96b431142262d47d727f309e37) for a reference.
  - [x] The added dependencies should be imported at runtime when the new connector is invoked, rather than as top-level imports.
- [x] Honors the conventions of `BaseConnectorConfig` defined in [unstructured/ingest/interfaces.py](unstructured/ingest/interfaces.py) which is passed through [the CLI](unstructured/ingest/main.py):
  - [x] If running with an `.output_dir` where structured outputs already exists for a given file, the file content is not re-downloaded from the data source nor is it reprocessed. This is made possible by implementing the call to `MyIngestDoc.has_output()` which is invoked in [MainProcess._filter_docs_with_outputs](ingest-prep-for-many/unstructured/ingest/main.py).
  - [x] Unless `.reprocess` is `True`, then documents are always reprocessed.
  - [x] If `.preserve_download` is `True`, documents downloaded to `.download_dir` are not removed after processing.
  - [x] Else if `.preserve_download` is `False`, documents downloaded to `.download_dir` are removed after they are **successfully** processed during the invocation of `MyIngestDoc.cleanup_file()` in [process_document](unstructured/ingest/doc_processor/generalized.py)
  - [x] Does not re-download documents to `.download_dir` if `.re_download` is False, enforced in `MyIngestDoc.get_file()`
  - [x] Prints more details if `.verbose` similar to [unstructured/ingest/connector/s3_connector.py](unstructured/ingest/connector/s3_connector.py).

### Note: This checklist is not completely filled
I want to be quite clear: I have not created a test file. There are two main reasons for this:
1. Using this connector requires a Reddit Client ID and Client Secret. This information is linked to a specific Reddit account, which should be owned and controlled by Unstructured-IO, rather than by a contributor like myself. As a result, I cannot reasonably set up a test script as it requires this client ID and secret information.
2. The output of this ingest connector will differ over time: Results from a search query will differ, and results from the "Hot" section of a subreddit will change over time. As a result, I cannot create an expected output for this connector.

Adding tests for this is possible, but a lot of work. A consideration is that any testing script cannot directly contain the ID and Secret, as users could steal this authentication information. This requires the use of environment variables, which means the test suite only passes for those within the company who have access to the required authentication.

Let me know how you feel about this.

### Installation
I've added `praw` as a dependency, installable via `pip install unstructured["reddit"]`. I ran `pip-compile` to add an `requirements/ingest-reddit.txt` file.

### CLI Help
The usage of the `unstructured-ingest` CLI is now like so:
```
Usage: unstructured-ingest [OPTIONS]

Options:
  --s3-url TEXT                   Prefix of s3 objects (files) to download.
                                  E.g. s3://bucket1/path/. This value may also
                                  be a single file.
  --s3-anonymous                  Connect to s3 without local AWS credentials.
  --subreddit-name TEXT           The name of a subreddit, without the "r\",
                                  e.g. "machinelearning"
  --reddit-client-id TEXT         The client ID, see https://praw.readthedocs.
                                  io/en/stable/getting_started/quick_start.htm
                                  l#prerequisites for more information.
  --reddit-client-secret TEXT     The client secret, see https://praw.readthed
                                  ocs.io/en/stable/getting_started/quick_start
                                  .html#prerequisites for more information.
  --reddit-user-agent TEXT        The user agent to use on the Reddit API, see
                                  https://praw.readthedocs.io/en/stable/gettin
                                  g_started/quick_start.html#prerequisites for
                                  more information.
  --reddit-search-query TEXT      If set, return posts using this query.
                                  Otherwise, use hot posts.
  --reddit-num-posts INTEGER      The number of posts to fetch.
  --re-download / --no-re-download
                                  Re-download files from s3 even if they are
                                  already present in --download-dir.
  --download-dir TEXT             Where s3 files are downloaded to, defaults
                                  to tmp-ingest-<6 random chars>.
  --preserve-downloads            Preserve downloaded s3 files. Otherwise each
                                  file is removed after being processed
                                  successfully.
  --structured-output-dir TEXT    Where to place structured output .json
                                  files.
  --reprocess                     Reprocess a downloaded file from s3 even if
                                  the relevant structured output .json file in
                                  --structured-output-dir already exists.
  --num-processes INTEGER         Number of parallel processes to process docs
                                  in.  [default: 2]
  -v, --verbose
  --help                          Show this message and exit.
```
Note the six new reddit-related arguments. The first three, `subreddit-name`, `reddit-client-id` and `reddit-client-secret` are mandatory if a user wishes to load from Reddit. See [this documentation](https://praw.readthedocs.io/en/stable/getting_started/quick_start.html#prerequisites) for details on how to set up a Client ID and secret. The other three allow users more control over the user agent that the Reddit API receives, the posts that are fetched, and the number of fetched posts.

### Sample usage
An example usage would be the following:
```
unstructured-ingest \
    --subreddit-name machinelearning \
    --reddit-client-id <client id here> \
    --reddit-client-secret <client secret here> \
    --reddit-user-agent "Unstructured Ingest Subreddit fetcher by \u\..." \
    --reddit-search-query "Unstructured" \
    --reddit-num-posts 10 \
    --structured-output-dir reddit-ingest-output \
    --num-processes 2 \
    --verbose
```

### Sample output
When running the above script, one of the files I receive is:

```json
[
  {
    "element_id": "bf23127adf48125ed8bd1b56c817b4a0",
    "text": "# [N] CfP: Workshop on Behavior-driven Autonomous Driving in Unstructured Environments at IROS 2022 in Kyoto, Japan on 27th October 2022.",
    "type": "UncategorizedText",
    "metadata": {}
  },
  {
    "element_id": "dafe7694460f4e37b708f5134f2f7f75",
    "text": "Hello,",
    "type": "UncategorizedText",
    "metadata": {}
  },
  {
    "element_id": "d93ebaca379f756d1bbcfc50f2046033",
    "text": "We are accepting **short (4+n) or long paper (8+n)** contributions to our workshop, \"Behavior-driven Autonomous Driving in Unstructured Environments (BADUE22)\", [removed manually for this PR].",
    "type": "NarrativeText",
    "metadata": {}
  },
  {
    "element_id": "4b9f854ca00cc8c10bbf85f23f167bab",
    "text": "*Deadline: Sept. 20, 2022 (AoE).**",
    "type": "ListItem",
    "metadata": {}
  },
  {
    "element_id": "118ee6ecc333e3d426ee676c35b65b3e",
    "text": "*Website:** [**https://gamma.umd.edu/workshops/badue22/**](https://gamma.umd.edu/workshops/badue22/)",
    "type": "ListItem",
    "metadata": {}
  },
  {
    "element_id": "abb88f7536c7e46d84c2f31cbcb1391e",
    "text": "*Submit:** [**https://cmt3.research.microsoft.com/BADUE2022/Submission/Index/**](https://cmt3.research.microsoft.com/BADUE2022/Submission/Index/)",
    "type": "ListItem",
    "metadata": {}
  },
  {
    "element_id": "fd4fc5dea175aac5466b26665f6f4079",
    "text": "*Contact organizer via email:** **rchandra@utexas.edu**",
    "type": "ListItem",
    "metadata": {}
  },
  {
    "element_id": "41868d23593094d3c89684955a663332",
    "text": "Autonomous driving research has advanced significantly through continuous improvements in perception, prediction, and planning. Despite this success, autonomous driving is limited to simple, sparse, [removed manually for this PR]",
    "type": "NarrativeText",
    "metadata": {}
  },
...
```

## Implementation details
The output of the files is formatted like so:
```
# {title}
{text_body}
```
And are written as `.md` files for further processing. Note that these files are now often processed as `text\plain` files, but #284 should add more proper support for Markdown, allowing the title to be properly parsed as a title (i.e. by taking advantage of the `#`).

Note that images are not handled by this, as the requirement by #272 states that the data must be stored in Markdown for processing.

Additionally, in PRAW it is only possible to fetch the posts in full in bulk, i.e. we cannot load X url's which we assign to X `RedditIngestDoc` instances, where each instance is responsible for loading only that URL. This is not an issue, however.

---

Let me know if you need anything of me at this point! Also, notify me if I need to resolve any merge conflicts: I'll be ready to help out.

- Tom Aarsen

### CODE REVIEW

**Comment by tomaarsen** [comment]
One of the CI jobs failed as it was unable to set up the job. This is unrelated to this PR, for which all tests pass, and which is ready for review. See the logs for the failing job here:
https://github.com/Unstructured-IO/unstructured/actions/runs/4276925608/jobs/7445412438
I'll push an empty commit to rerun the CI.

File: Makefile [hunk 1]
```diff
@@ -54,6 +54,10 @@ install-build:
 install-ingest-s3:
 	pip install -r requirements/ingest-s3.txt
 
+.PHONY: install-ingest-github
+install-ingest-github:
```
[the remark this answers is not in the document]
Comment (cragwolfe) [reply Makefile:58]: please add to pip-compile as well

**Review by cragwolfe — APPROVED** [review]
LGTM conditional on Makefile fixes.

Tested the the `examples/ingest/reddit/ingest.sh`. Wonderful!

File: unstructured/ingest/interfaces.py [hunk 2]
```diff
@@ -90,7 +90,7 @@ def write_result(self):
     def process_file(self):
         print(f"Processing {self.filename}")
 
-        elements = partition(filename=self.filename)
+        elements = partition(filename=str(self.filename))
```
[the remark this answers is not in the document]
Comment (tomaarsen) [reply unstructured/ingest/interfaces.py:93]: Indeed. libmagic won't accept the Path object.

File: Makefile (hunk 1 above)
[the remark this answers is not in the document]
Comment (tomaarsen) [reply Makefile:58]: Oops, embarrassing! As for `pip-compile`, does that not currently fail due to `requirements/ingest-s3.in` not existing?
That said, I've added `pip-compile --upgrade --extra=reddit --output-file=requirements/ingest-reddit.txt requirements/base.txt setup.py`. Let me know if that's alright!

I think the S3-line in the Makefile might need to be:
`pip-compile --upgrade --extra=s3 --output-file=requirements/ingest-s3.txt requirements/base.txt setup.py`

[the remark this answers is not in the document]
Comment (cragwolfe) [reply Makefile]: The additions look good! 🙏 

Yes, the s3 pip-compile is broken 😳

**Pull request merged**